### ISSUE TITLE
ttnn.maximum / ttnn.minimum / ttnn.clamp / ttnn.hardtanh silently drop NaN — a single SFPSWAP is a permutation, so max and min can never both return NaN

### ISSUE THREAD

**Issue by bvdyvncqr5-commits** [tenstorrent/tt-metal#51470]
# ttnn.maximum / ttnn.minimum / ttnn.clamp / ttnn.hardtanh silently drop NaN — a single SFPSWAP is a permutation, so max and min can never both return NaN

`ttnn.maximum`, `ttnn.minimum`, `ttnn.clamp` and `ttnn.hardtanh` are each implemented by one or two bare `SFPSWAP` instructions. `SFPSWAP` is a compare-and-exchange, so its two output registers always hold a **permutation** of its two inputs. PyTorch requires `torch.maximum(NaN, x)` **and** `torch.minimum(NaN, x)` to *both* return NaN — a permutation of `{NaN, x}` can never be `(NaN, NaN)`.

This is a purely structural argument: **at least one of `ttnn.maximum` / `ttnn.minimum` is wrong for every NaN input, on both Wormhole and Blackhole, regardless of how the hardware orders NaN.**

Concretely, NaN is silently replaced by the other operand:

* `ttnn.minimum(NaN, 5.0)` returns `5.0`
* `ttnn.maximum(-NaN, 5.0)` returns `5.0`
* `ttnn.clamp(NaN, -2.0, 2.0)` returns `2.0`

The int32 variant in the same kernel already applies a correction for `SFPSWAP` ordering, so there is a ready structural precedent for what the float32 path would need.

### Summary

`SFPSWAP` is the hardware compare-and-exchange used for min/max. Three in-repo facts establish that it is a permutation:

1. `sfpi_lib.h:428` `min_max()` issues a single `__builtin_rvtt_sfpswap` and then uses `sfpselect2` to pick one of *that one instruction's* two outputs.
2. `swap(a, b)` in the same header is `SFPSWAP_MOD1_SWAP` — the instruction's semantics are literally "exchange or don't".
3. `ckernel_sfpu_binary_max_min.h:78` contains the comment *"Conditionally swap again to fix the cases where SFPSWAP got the result backwards"* — only a swap network makes "got it backwards, swap again" meaningful.

`calculate_binary_max_min` issues one `TTI_SFPSWAP(0, LREG1, LREG0, SFPSWAP_MOD1_VEC_MIN_MAX)` and then stores `LREG1` for max or `LREG0` for min. The `SFPLOADMACRO` path routes through `binary_max_min_init`'s `InstructionTemplate[0]`, which is the same single `SFPSWAP` (`mod1 = 9` for max, `1` for min). Either way, the two candidate results are a permutation of the two inputs.

Since `torch.maximum` and `torch.minimum` must *both* yield NaN when either operand is NaN, and a permutation of `{NaN, x}` contains exactly one NaN, **one of the two ops is necessarily wrong on every NaN input.** No assumption about the hardware's NaN ordering is needed to reach that conclusion.

Which one is wrong follows from the ordering `SFPSWAP` actually uses. Three in-repo comments state it is sign-magnitude:

* `ckernel_sfpu_unary_max_min.h:80`: *"it expects sign-magnitude integers"*
* `ckernel_sfpu_relu.h:24`: *"the hardware sign-magnitude min/max (SFPSWAP)"*, and `:85`: *"0x80000000 is -0"*

(Both of those comments sit in int32 code paths, but `SFPSWAP`'s comparison
semantics are a property of the instruction, not of the caller — and for IEEE
float32 the sign-magnitude order coincides with the real order on all finite
values and ±inf, which is precisely why the instruction is usable for float
min/max at all. Note also that **the permutation argument above does not depend
on this at all** — the ordering only determines *which* of `maximum`/`minimum`
is wrong for a given NaN sign, not *that* one of them must be.)

For IEEE float32, sign-magnitude integer order agrees with real order on all finite values and ±inf. Extending it to NaN puts `|NaN|` (`0x7FC00000`) above `|inf|` (`0x7F800000`), so `+NaN` sorts highest and `-NaN` sorts lowest. Hence `minimum(+NaN, x) = x` and `maximum(-NaN, x) = x`.

- Files:
  - `tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_max_min.h` (WH)
  - `tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary_max_min.h` (BH, logic-identical)
  - `ckernel_sfpu_unary_max_min.h`, `ckernel_sfpu_clamp.h`, `ckernel_sfpu_hardtanh.h` (both platforms)
- Fix template, already in-repo: `ckernel_sfpu_comp.h` — see below.

### Measured behaviour

Simulated per-instruction in float32 against `torch` / `numpy` reference. Errors are semantic (a non-NaN value escapes where NaN is required), so relative error is undefined rather than large.

| input | returned | expected |
|---|---|---|
| `ttnn.minimum(+NaN 0x7fc00000, 5.0)` | `5.0` | `nan (0x7fc00000)` |
| `ttnn.minimum(+NaN, -inf)` | `-inf` | `nan (0x7fc00000)` |
| `ttnn.minimum(+NaN, 0.0)` | `0.0` | `nan (0x7fc00000)` |
| `ttnn.maximum(-NaN 0xffc00000, 5.0)` | `5.0` | `nan (0x7fc00000)` |
| `ttnn.maximum(-NaN, +inf)` | `inf` | `nan (0x7fc00000)` |
| `ttnn.clamp(+NaN, min=-2.0, max=2.0)` | `2.0` | `nan (0x7fc00000)` |
| `ttnn.clamp(+NaN, min=0.0, max=6.0)` | `6.0` | `nan (0x7fc00000)` |
| `ttnn.clamp(-NaN, min=0.0, max=6.0)` | `0.0` | `nan (0xffc00000)` |
| `ttnn.hardtanh(+NaN, min_val=-1.0, max_val=1.0)` | `1.0` | `nan` |

As a control, 200,100 random and special **non-NaN** pairs (including ±inf, ±0, subnormals) were run through the same model: **0 mismatches** for both `maximum` and `minimum`. The defect is strictly NaN-specific — this is a semantics bug, not an accuracy claim, so there is no tolerance threshold to argue about.

One detail worth noting for whoever writes the fix: the expected bit patterns above were measured against torch directly, and the two operator families do not agree on the NaN sign. `torch.maximum` / `torch.minimum` normalise the result to `0x7fc00000` regardless of which operand was NaN or what its sign was, whereas `torch.clamp` and `torch.nn.functional.hardtanh` preserve the input's sign bit (`-NaN` in, `0xffc00000` out). A fix for `max`/`min` is therefore free to emit any quiet NaN, but a fix for `clamp`/`hardtanh` should pass the input through rather than substitute a canonical NaN.

### clamp and hardtanh: same root cause, second operator family

`calculate_clamp` is `max` then `min`:

```
load_value_param_float(min_val);  calculate_unary_max_min_float_body<Max>();
load_value_param_float(max_val);  calculate_unary_max_min_float_body<Min>();
```

and each body is a single `TTI_SFPSWAP(0, LREG12, LREG0, ...)`. `hardtanh` calls `sfpi::clamp`, which is `min(max(val, lower), upper)` — the same two swaps.

For NaN to survive, it would have to compare **greater** than `lo` in step one (so `max` keeps it) and **less** than `hi` in step two (so `min` keeps it). A comparator producing a consistent total order cannot do both, so under *any* total ordering of NaN — highest, lowest, or anywhere in between — clamp loses it. Under sign-magnitude order specifically, `+NaN → hi` and `-NaN → lo`.

One further possibility deserves stating explicitly, since ruling it out is what makes the above exhaustive: a comparator could treat NaN as **unordered** and therefore never swap at all. NaN would then survive both steps and `clamp` would be correct. That case is excluded by the same in-tree evidence cited earlier — `SFPSWAP` compares its operands as sign-magnitude **integers**, and integer comparison has no unordered case; every bit pattern has a definite position. If that reading turned out to be wrong, this clamp/hardtanh section would weaken accordingly. **The `maximum`/`minimum` argument would not**: two output slots holding a permutation of two inputs contain exactly one NaN no matter what the comparator decides, ordered or unordered.

### Test gap

`ttnn.maximum` and `ttnn.minimum` do have dedicated coverage — `test_binary_maximum.py` and `test_unary_maximum.py`, plus four sweeps under `sweeps/eltwise/composite/binary/maximum/`. Those tests exercise float32, and they explicitly parametrise special values: `±inf`, `±FLT_MAX`, and the int32 extremes appear across roughly ten cases between the two files.

**Not one of them uses NaN.** Special-value coverage was clearly considered and deliberately built out; NaN is the one class of special value missing from it.

The same gap shows against a standard the repository already applies elsewhere: `test_binary_comp_fp32.py` defines a `_SPECIAL_FP32_GRID` that *does* contain `nan` alongside `±inf`, and uses it to test the tensor–tensor paths of `ttnn.eq/ne/gt/lt/ge/le`.

**The standard already exists** — order-relation operators are expected to be NaN-correct, and there is a special-value grid built to enforce it. `max`/`min` are order relations by the same definition but are not currently covered by that grid, so extending it is a natural fit.

For clamp, `test_backward_clamp.py` is backward-only, all bfloat16, with inputs from `data_gen_with_range(shape, -100, 100)` (finite only) and a min/max parameter list containing no inf and no NaN. There is no forward clamp test file. `test_hardtanh.py` uses `torch.randn(...).bfloat16() * 10` — no NaN, no inf.

### Proposed fix

Detect NaN operands and force a NaN result, before or after the swap. `ckernel_sfpu_comp.h` already contains the idiom this needs — its comment at `:74` reads *"rejects NaN with a single SFPIADD"*, implemented at `:113` as:

```cpp
TTI_SFPIADD(0, INF, ABS_V, sfpi::SFPIADD_MOD1_ARG_2SCOMP_LREG_DST | sfpi::SFPIADD_MOD1_CC_GTE0);
```

i.e. compare `|v|` against the inf bit pattern as integers; NaN is exactly the case `|v| > 0x7F800000`. This is one instruction per operand and needs no new constants.

The int32 variant in `ckernel_sfpu_binary_max_min.h` already demonstrates the structural pattern for correcting `SFPSWAP` — `SFPSETCC` on both registers, a second conditional `SFPSWAP`, then `SFPENCC`. The float32 path needs the analogous guard, keyed on the NaN test above rather than on sign.

For `clamp` / `hardtanh`, one guard on the input value before the two swaps is sufficient, since the bounds are compile-time scalars.

Note the `SFPLOADMACRO` fast paths (`binary_max_min_init`, `calculate_unary_max_min`) will need the same treatment as the `DISABLE_SFPLOADMACRO` paths, or the two will diverge.

### Success criteria

- [ ] `ttnn.maximum` / `ttnn.minimum` propagate NaN for both `+NaN` and `-NaN`, matching `torch.maximum` / `torch.minimum`
- [ ] `ttnn.clamp` / `ttnn.hardtanh` return NaN for NaN input, matching `torch.clamp` / `torch.nn.functional.hardtanh`
- [ ] Both the `SFPLOADMACRO` and `DISABLE_SFPLOADMACRO` paths are fixed and agree
- [ ] `max`/`min`/`clamp`/`hardtanh` added to the existing `_SPECIAL_FP32_GRID` coverage in `test_binary_comp_fp32.py`, or an equivalent special-value grid
- [ ] Non-NaN behaviour unchanged (verified here on 200,100 pairs: 0 mismatches)
- [ ] Confirmed on Blackhole (logic-identical file — same fix applies)

### Why this matters

NaN propagation is the standard signal that a computation has gone wrong. Silently substituting a finite operand converts a detectable failure into corrupted data that flows downstream unnoticed — `clamp` is especially exposed, since it turns NaN into a *plausible-looking* bound value.

These are public ttnn ops with direct torch counterparts whose NaN behaviour is explicitly specified. The repository already treats NaN correctness as a requirement for comparison operators and has both the test grid and the detection idiom in place; `max`, `min`, `clamp` and `hardtanh` are order relations that are not yet covered by either, and both pieces needed to cover them already exist in-tree.

The argument that at least one of `maximum`/`minimum` must be wrong needs no hardware measurement — it follows from `SFPSWAP` being a permutation, which the repository's own source and comments establish.

Happy to open a PR with the kernel change and a pytest extending the existing special-value grid to `max`/`min`/`clamp`/`hardtanh`, if that would be useful.

**Comment by bvdyvncqr5-commits** [comment]
Filling in the bug report template fields. This one is different from my other reports: it does not rest on a measurement, because it does not need one.

### Component / Area

kernels (eltwise binary)

### Issue Type (optional)

Bad Outputs

### Observed

`ttnn.maximum` / `ttnn.minimum` on the float path do not propagate NaN.

### Expected

NaN in either operand produces NaN out, for **both** `maximum` and `minimum` (torch semantics).

### Why no reproduction is attached

`calculate_binary_max_min` produces its result with a single `TTI_SFPSWAP`
(`ckernel_sfpu_binary_max_min.h:29`). SFPSWAP is a compare-and-swap: its two output slots are always
a permutation of its two inputs. For NaN to survive, the same comparator would have to rank it as the
larger operand for `maximum` and as the smaller one for `minimum`. One comparator cannot do both, so
for every NaN input at least one of the two operations returns the wrong value — whatever order the
hardware happens to impose on NaN.

I would rather state it this way than attach a number, because the conclusion does not depend on one.

### The same file already carries the fix — for int32

Lines 76-82 of the same file follow the swap with `SFPSETCC` x2, a second `SFPSWAP` and `SFPENCC`,
under this comment:

> `// Conditionally swap again to fix the cases where SFPSWAP got the result backwards`

The float32 path at line 29 has no equivalent.

### 1. Steps (exact commands)

**I have not run this.** I have no Tenstorrent hardware, and on the bfloat16 path NaN is flattened
before I can observe it. For anyone who can run it, one command settles it:

```python
nan = torch.full((32, 32), float("nan"), dtype=torch.float32)
five = torch.full((32, 32), 5.0, dtype=torch.float32)
a = ttnn.from_torch(nan, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
b = ttnn.from_torch(five, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)

print(ttnn.to_torch(ttnn.maximum(a, b))[0, 0])   # torch.maximum -> nan
print(ttnn.to_torch(ttnn.minimum(a, b))[0, 0])   # torch.minimum -> nan
```

At least one of the two must disagree with torch.

### 2. Input data / link or description

None beyond the two constants above.

### 3. Frequency

Always, by construction — not sampling-dependent.

### 1. Software Versions

Not applicable; the argument is from the source in this repository at `main`.

### 2. Hardware Details

None — I do not have Tenstorrent hardware.

### Is this a regression?

Unknown.

---

*Disclosure: the analysis in this comment was produced with AI assistance. I have reviewed it and take responsibility for it. It rests entirely on the two source locations cited above, so it can be checked without taking my word for anything.*
